Source/Applications/openHistorian/openHistorian/ServiceHost.cs

```diff
@@ -933,6 +942,180 @@ private void SetupGrafanaHostingAdapter()
+                            break;
+                        case "secondary":
+                            enableFailover = true;
+                            break;
+                        default:
+                            LogStatusMessage($"Encountered invalid failover 'NodeType' setting \"{nodeType}\". Disabling failover node.", UpdateType.Warning);
+                            break;
+                    }
+                }
+
+                if (enableFailover)
+                {
+                    failoverSettings.Add("FailoverEndpoint", DefaultFailoverURL, "Defines the URL of the other node(s) in the failover system. This may include multiple URLs separated by semi-colons.");
+                    failoverSettings.Add("RestartDelay", DefaultFailoverRestartDelay, "Defines the delay, in milliseconds, before the node will attempt to start again.");
+
+                    if (string.IsNullOrWhiteSpace(failoverSettings["RestartDelay"].Value) || !int.TryParse(failoverSettings["RestartDelay"].Value, out delay))
+                        delay = DefaultFailoverRestartDelay;
+                    
+                    nodes = string.IsNullOrWhiteSpace(failoverSettings["FailoverEndpoint"].Value) ?
+                        [ DefaultFailoverURL ] : 
+                        failoverSettings["FailoverEndpoint"].Value.Split(';');
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.SwallowException(ex);
+                return false;
+            }
+
+            if (!enableFailover)
+                return false;
+
+            LogStatusMessage("Failover mode enabled, starting node polling...");
+            
+            foreach (string node in nodes)
+            {
+                string[] settings = node.Split('&');
+                string url = settings[0];
+
+                if (!PingNode(url))
+                    continue;
+                
+                if (isPrimary)
+                {
+                    if (settings.Length == 1)
+                    {
+                        LogStatusMessage($"No host specified for \"{url}\" - system will not be able to shut down the node", UpdateType.Alarm);
+                        continue;
+                    }
+
+                    string remoteHost = settings[1];
+                    
+                    LogStatusMessage($"Failover logic detected: \"{url}\" is active - stopping node on {remoteHost}");
+                    StopRemoteService(remoteHost);
+                    
+                    // if no response override Connection Log Files with Failover Log Files
+                    throw new NotImplementedException();
+                }
+                   
+                LogStatusMessage($"Failover logic detected: \"{url}\" is active - restarting in {delay:N0} seconds.", UpdateType.Warning);
+                
+                // Update failover cache file
+                DelayedRestart(delay);
```

Comment:
This restarts the service whether the service is primary or secondary - should this be in an `else` condition or is this desired behavior to retry remote operations?

Reply:
Intend was to `continue` if it is Primary. That has been added,